File: tests/contract.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import { startFixture } from './preview.mjs';

// The same sequence is exercised through the actual Setec provider in Go.
test('PROVIDER-1: UI fixture conforms to shared Setec lifecycle cases', async () => {
  const cases = JSON.parse(await readFile(new URL('./setec-contract.json', import.meta.url)));
  const fixture = await startFixture(0);
  try {
    fixture.secrets.clear();
    fixture.secrets.set('finance/hidden', {Name: 'finance/hidden', ActiveVersion: 1, Values: {1: 'eA=='}});
    fixture.secrets.set('finance/private/password', {Name: 'finance/private/password', ActiveVersion: 1, Values: {1: 'eA=='}});
    const {connection} = await (await fetch(fixture.url + '/ui-api/session')).json();
    for (const tc of cases) {
      const res = await fetch(fixture.url + '/ui-api/' + tc.op, {
        method: 'POST', headers: {'Content-Type': 'application/json', 'X-TailVault-Connection': connection},
        body: JSON.stringify(tc.body),
      });
      assert.equal(res.status, tc.status, tc.label);
      if (Object.hasOwn(tc, 'result')) assert.deepEqual(await res.json(), tc.result, tc.label);
      else await res.text();
    }
  } finally { await fixture.close(); }
});

File: tests/provider/contract_test.go
package provider_test

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http/httptest"
	"os"
	"reflect"
	"testing"

	"github.com/angerops/tailvault/internal/portal"
	appRuntime "github.com/angerops/tailvault/internal/runtime"
	"github.com/tailscale/setec/server"
	"github.com/tailscale/setec/setectest"
	"tailscale.com/client/tailscale/apitype"
	"tailscale.com/tailcfg"
)

// These same cases exercise tests/preview.mjs. Only the network and WhoIs are
// substituted: all requests pass through the real portal, transport and Setec DB.
func TestProviderContract(t *testing.T) {
	raw, err := os.ReadFile("../setec-contract.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []struct {
		Label  string
		Op     string
		Body   json.RawMessage
		Status int
		Result json.RawMessage
	}
	if err := json.Unmarshal(raw, &cases); err != nil {
		t.Fatal(err)
	}
	db := setectest.NewDB(t, nil)
	db.MustPut(db.Superuser, "finance/hidden", "synthetic denied value")
	db.MustPut(db.Superuser, "finance/private/password", "x")
	ss := setectest.NewServer(t, db, &setectest.ServerOptions{WhoIs: func(ctx context.Context, addr string) (*apitype.WhoIsResponse, error) {
		who, err := setectest.AllAccess(ctx, addr)
		// Permission is decided by the actual server, including filtering list.
		who.CapMap[server.ACLCap] = []tailcfg.RawMessage{
			`{"action":["info","get","put","create-version","activate","delete"],"secret":["personal/*"]}`,
			`{"action":["info"],"secret":["finance/private/*"]}`,
		}
		return who, err
	}})
	hs := httptest.NewServer(ss.Mux)
	defer hs.Close()
	who := portal.Identity{UserID: 1, NodeID: "n-fixture", IP: "127.0.0.1", Login: "synthetic@example.test"}
	app, err := portal.New(portal.Config{
		Server:   "https://secrets.example.ts.net",
		Identity: func(context.Context) (portal.Identity, error) { return who, nil },
		Command: func(ctx context.Context, identity portal.Identity, _, op string, body []byte) ([]byte, int, error) {
			return appRuntime.Command(ctx, identity, hs.URL, op, body)
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	defer app.Close()
	token := ""
	request := func(method, op string, body []byte) *httptest.ResponseRecorder {
		req := httptest.NewRequest(method, "http://wails/ui-api/"+op, bytes.NewReader(body))
		req.Header.Set("Origin", "wails://wails")
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("X-TailVault-Connection", token)
		res := httptest.NewRecorder()
		app.ServeHTTP(res, req)
		return res
	}
	res := request("GET", "session", nil)
	var session struct{ Connection string }
	if res.Code != 200 || json.Unmarshal(res.Body.Bytes(), &session) != nil || session.Connection == "" {
		t.Fatal("cannot open synthetic view")
	}
	token = session.Connection
	for _, tc := range cases {
		t.Run(tc.Label, func(t *testing.T) {
			res := request("POST", tc.Op, tc.Body)
			if res.Code != tc.Status {
				t.Fatalf("status %d, want %d: %s", res.Code, tc.Status, res.Body.String())
			}
			if tc.Result != nil {
				var got, want any
				if err := json.Unmarshal(res.Body.Bytes(), &got); err != nil {
					t.Fatal(err)
				}
				if err := json.Unmarshal(tc.Result, &want); err != nil {
					t.Fatal(err)
				}
				if !reflect.DeepEqual(got, want) {
					t.Fatalf("got %s; want %s", res.Body.Bytes(), tc.Result)
				}
			}
		})
	}
}
